api/v1alpha1/backend_types.go

```diff
@@ -0,0 +1,153 @@
+//
+// +kubebuilder:validation:XValidation:rule="(has(self.fqdn) || has(self.ip) || has(self.unix))",message="one of fqdn, ip or unix must be specified"
+// +kubebuilder:validation:XValidation:rule="((has(self.fqdn) && !(has(self.ip) || has(self.unix))) || (has(self.ip) && !(has(self.fqdn) || has(self.unix))) || (has(self.unix) && !(has(self.ip) || has(self.fqdn))))",message="only one of fqdn, ip or unix can be specified"
+// +notImplementedHide
+type BackendAddress struct {
+	// FQDN defines a FQDN address
+	//
+	// +optional
+	FQDN *FQDNAddress `json:"fqdn,omitempty"`
+
+	// IP defines a IPv4 address
+	//
+	// +optional
+	IP *IPAddress `json:"ip,omitempty"`
+
+	// Unix defines the unix domain socket path
+	//
+	// +optional
+	Unix *UnixSocket `json:"unix,omitempty"`
+}
+
+// IPAddress describes TCP/UDP socket address, corresponding to Envoy's Socket Address
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-socketaddress
+// +notImplementedHide
+type IPAddress struct {
+	// Host defines to the IPv4 address of the backend service.
+	//
+	// +kubebuilder:validation:MinLength=7
+	// +kubebuilder:validation:MaxLength=15
+	// +kubebuilder:validation:Pattern=`^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$`
+	Host string `json:"host"`
+
+	// Port defines to the port of of the backend service.
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65535
+	Port int32 `json:"port"`
+}
+
+// FQDNAddress describes TCP/UDP socket address, corresponding to Envoy's Socket Address
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-socketaddress
+// +notImplementedHide
+type FQDNAddress struct {
+	// Host defines to the FQDN address of the backend service.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=`^(\*\.)?[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
+	Host string `json:"host"`
+
+	// Port defines to the port of of the backend service.
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65535
+	Port int32 `json:"port"`
+}
+
+// UnixSocket describes TCP/UDP unix domain socket address, corresponding to Envoy's Pipe
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-pipe
+// +notImplementedHide
+type UnixSocket struct {
+	Path string `json:"path"`
+}
+
+// BackendSpec describes the desired state of BackendSpec.
+// +notImplementedHide
+type BackendSpec struct {
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=4
+	// +kubebuilder:validation:XValidation:rule="self.all(f, has(f.fqdn)) || !self.exists(f, has(f.fqdn))",message="fqdn addresses cannot be mixed with other address types"
+	BackendAddresses []BackendAddress `json:"addresses,omitempty"`
```

Comment:
imo `backend` prefix is not needed, the config looks like
```
backend:
  spec:
    backendAddresses
```

this could be simplified to
```
backend:
  spec:
    addresses:

Reply:
Sorry, missed your comment. I'll rename to endpoints as you suggested.

Reply:
tbh I dont know which one is better - `endpoints` or `addresses`

Reply:
I think that your suggestion is closer to both K8s Endpoint and Envoy Endpoint APIs, where endpoints have an address (or addresses). So, I think that it should be pretty intuitive to EG users.